### ISSUE TITLE
[FEAT][NORM_SOFTMAX] infer normalization and softmax shape metadata from inputs

### ISSUE THREAD

**Issue by superAngGao** [tile-ai/TileOPs#1686]
Parent tracking issue: #1674

### Symptom / Motivation
Upstream issue #1648 established the API principle that public Op constructors should not require users to repeat shape metadata that is directly derivable from forward input tensor metadata. #1673 applied that principle to Conv and AvgPool. This issue applies the same cleanup to the spatial / batch normalization and softmax-family Ops that still require shape metadata in their constructors.

The current spatial normalization constructors require values such as `N`, `C`, `spatial`, and `dtype`, even though those are available from the tensors passed to `forward()`:

- `BatchNormFwdOp(N, C, spatial=..., dtype=..., ...)`
- `BatchNormBwdOp(N, C, *spatial, dtype=..., ...)`
- `GroupNormFwdOp(N, C, spatial, num_groups, dtype, ...)`
- `GroupNormFwdOpNoAffine(N, C, spatial, num_groups, dtype, ...)`
- `InstanceNormFwdOp(N, C, spatial, dtype, ...)`
- `InstanceNormFwdOpNoAffine(N, C, spatial, dtype, ...)`

This makes the preferred API more verbose than necessary and can let constructor-provided dimensions drift from the actual tensors passed to `forward()`.

The softmax-family wrappers have a narrower version of the same problem:

- `SoftmaxFwdOp(N, dtype, dim=...)`
- `LogSoftmaxFwdOp(N, dtype, dim=...)`

Their shared `_SoftmaxBaseOp` already performs forward-time `M/N` derivation and lazy kernel lookup, but the public wrappers still require callers to pass `N` and `dtype` up front.

This issue intentionally focuses on spatial / batch normalization. Row-wise and layer-style normalization cleanup is already covered separately or has already moved toward input-inferred/lazy specialization patterns:

- `RMSNormFwdOp`
- `LayerNormFwdOp`
- `AdaLayerNormFwdOp`
- `AdaLayerNormZeroFwdOp`
- `FusedAddLayerNormFwdOp`
- `FusedAddRMSNormFwdOp`

### Root Cause Analysis
The current BatchNorm, GroupNorm, and InstanceNorm Ops build shape-specialized kernels eagerly in `__init__()`, so the constructors require all dimensions needed to instantiate the kernel before any tensor is seen. `forward()` then validates actual tensor shapes against constructor-committed shapes instead of deriving the specialization from the actual inputs.

For `SoftmaxFwdOp` and `LogSoftmaxFwdOp`, the base implementation is already mostly lazy, but the subclass constructors still force `N` and `dtype` as required arguments. This leaves the public API inconsistent with the input-inferred contract and with other reduction Ops that bind dimensions during `forward()`.

For GroupNorm and InstanceNorm, the Op layer also pre-allocates device constants such as unit weights / zero bias buffers at construction time, which couples the Op instance to the active CUDA device and to the constructor-provided specialization. These buffers should be created lazily as part of the resolved specialization and cached consistently with the kernel.

### Related Files
- `tileops/ops/norm/batch_norm.py`: `BatchNormFwdOp`, `BatchNormBwdOp`
- `tileops/ops/norm/group_norm.py`: `GroupNormFwdOp`, `GroupNormFwdOpNoAffine`
- `tileops/ops/norm/instance_norm.py`: `InstanceNormFwdOp`, `InstanceNormFwdOpNoAffine`
- `tileops/ops/reduction/softmax.py`: `SoftmaxFwdOp`
- `tileops/ops/reduction/log_softmax.py`: `LogSoftmaxFwdOp`
- `tileops/ops/reduction/_softmax_base.py`
- `tileops/kernels/norm/batch_norm.py`
- `tileops/kernels/norm/group_norm.py`
- `tileops/kernels/norm/instance_norm.py` if present for no-affine paths
- `tileops/kernels/reduction/softmax.py`
- `tileops/manifest/normalization.yaml`
- `tileops/manifest/reduction.yaml`
- `tests/ops/test_batch_norm.py`
- `tests/ops/test_group_norm.py`
- `tests/ops/test_instance_norm.py`
- `tests/ops/test_softmax.py`
- `benchmarks/ops/bench_batch_norm.py`
- `benchmarks/ops/bench_group_norm.py`
- `benchmarks/ops/bench_instance_norm.py`
- `benchmarks/ops/bench_softmax.py`

## Goal
Align the spatial / batch normalization and softmax-family public APIs with the #1648 input-inferred shape metadata principle while preserving semantic/config parameters as explicit constructor arguments.

Preferred APIs should infer tensor-derived metadata in `forward()` and build kernels lazily:

- BatchNorm forward: infer `N`, `C`, spatial extents, and input dtype from `x`; validate `weight`, `bias`, `running_mean`, and `running_var` against inferred `C`.
- BatchNorm backward: infer `N`, `C`, spatial extents, and dtype from `grad_out` / `x`; validate `weight`, `mean`, and `rstd` against inferred `C`.
- GroupNorm affine: infer `N`, `C`, spatial extents, and dtype from `x`; validate `weight` and `bias` against inferred `C`.
- GroupNorm no-affine: infer `N`, `C`, spatial extents, and dtype from `x`.
- InstanceNorm affine: infer `N`, `C`, spatial extents, and dtype from `x`; validate `weight` and `bias` against inferred `C`.
- InstanceNorm no-affine: infer `N`, `C`, spatial extents, and dtype from `x`; validate running stats when the eval-mode path uses them.
- Softmax / LogSoftmax: infer `N`, `M`, dtype, and CUDA device from `x`; keep `dim` as the explicit semantic axis parameter.

Keep semantic/config parameters explicit:

- BatchNorm: `training`, `momentum`, `eps`
- GroupNorm: `num_groups`, `eps`
- InstanceNorm: `use_input_stats`, `momentum`, `eps`
- Softmax / LogSoftmax: `dim`
- Kernel override and tuning knobs: `kernel_map`, `tune`

## Plan

1. Update the scoped normalization Ops so preferred constructors no longer require `N`, `C`, `spatial`, or `dtype` when those values can be derived from forward tensors.
2. Replace eager kernel construction with lazy kernel caches keyed by the resolved specialization signature, including shape, dtype, CUDA device, affine/no-affine variant, training/inference path, `num_groups`, `eps`, and tuning-relevant config.
3. Move device-dependent pre-allocated constants in GroupNorm / InstanceNorm into the lazy specialization cache so a single Op instance is not accidentally bound to the CUDA device active at construction time.
4. Update `SoftmaxFwdOp` and `LogSoftmaxFwdOp` wrappers so preferred constructors no longer require `N` or `dtype`; reuse `_SoftmaxBaseOp` forward-time specialization.
5. Preserve practical compatibility paths where useful, but make the preferred API and all examples input-inferred.
6. Update `eval_roofline()` to use the active or most recent resolved specialization, or fail clearly if called before a specialization is available.
7. Update `tileops/manifest/normalization.yaml` and `tileops/manifest/reduction.yaml` so `signature.params`, `static_dims`, `shape_rules`, workloads, roofline bindings, and source metadata match the implemented preferred API.
8. Update unit tests to cover preferred input-inferred construction, legacy compatibility or migration behavior if retained, shape/dtype mismatch validation, lazy cache reuse, changing-shape behavior, and multi-device behavior where practical.
9. Update benchmarks and in-repo call sites to use the preferred input-inferred API unless they intentionally test compatibility.

## Kernel Attribute Contract
Follow the lazy-specialization style used by the closed Conv / Pool cleanup in #1673:

- `self._kernel_cache` is the authoritative cache of compiled/built kernel specializations.
- `self.kernel` may be retained for compatibility and debugging, but it must mean "the most recently used kernel", not "the only kernel owned by this Op".
- `self.kernel` should not be eagerly built in `__init__()`. Initialize it to `None` if the class exposes it, then assign it in `forward()` after resolving the specialization:

```python
kernel = self._get_kernel(resolved_specialization)
self.kernel = kernel
```

- If a later `forward()` call resolves to a different specialization, `self.kernel` should be overwritten with that specialization's cached kernel.
- A cache hit should not trigger JIT/build work; it should only do metadata validation, specialization-key construction, dictionary lookup, and the `self.kernel` last-used assignment.
- Core execution, roofline logic, tests, and benchmarks should rely on `_kernel_cache` / `_last_roofline_spec` rather than treating `self.kernel` as canonical state.

## Constraints
- Do not infer scheduling metadata by reading CUDA tensor values.
- Do not change normalization mathematical semantics relative to the existing PyTorch reference paths.
- Do not change softmax / log_softmax mathematical semantics or `dim=None` compatibility behavior relative to the existing PyTorch reference path.
- Keep row-wise/layer-style normalization cleanup out of this issue unless a follow-up explicitly expands the scope.
- Keep affine and no-affine paths distinct when the kernel contracts require different specializations.
- Preserve clear validation errors for rank, channel, affine parameter, running-stat, dtype, and device mismatches.
- Avoid binding an Op instance to a CUDA device at construction time when no tensor has been seen yet.

## Acceptance Criteria
- [ ] Preferred BatchNorm forward/backward APIs no longer require constructor arguments for `N`, `C`, `spatial`, or `dtype` derivable from forward tensors.
- [ ] Preferred GroupNorm affine/no-affine APIs no longer require constructor arguments for `N`, `C`, `spatial`, or `dtype` derivable from `x`.
- [ ] Preferred InstanceNorm affine/no-affine APIs no longer require constructor arguments for `N`, `C`, `spatial`, or `dtype` derivable from `x`.
- [ ] Preferred Softmax / LogSoftmax APIs no longer require constructor arguments for `N` or `dtype` derivable from `x`.
- [ ] Kernels and device-local helper buffers are lazily built and cached by the resolved specialization signature.
- [ ] `eval_roofline()` reports values for the resolved specialization and fails clearly if called before a specialization is available.
- [ ] Manifest entries for scoped normalization and softmax-family Ops match the implemented public contract, including params, shape rules, roofline metadata, and static dimension declarations.
- [ ] Unit tests cover correctness, validation, cache reuse, changing-shape behavior, and any retained compatibility behavior.
- [ ] Benchmarks and in-repo call sites use the preferred input-inferred API unless they are compatibility tests.
- [ ] Modified files pass targeted normalization / softmax tests and manifest validation.
